File: src/main/go/g0301_0400/s0338_counting_bits/solution.go
package s0338_counting_bits

// #Easy #Dynamic_Programming #Bit_Manipulation #LeetCode_75_Bit_Manipulation
// #Udemy_Bit_Manipulation #Big_O_Time_O(num)_Space_O(num)
// #2025_05_10_Time_0_ms_(100.00%)_Space_6.48_MB_(48.26%)

func countBits(num int) []int {
	result := make([]int, num+1)
	borderPos := 1
	incrPos := 1
	for i := 1; i < len(result); i++ {
		if incrPos == borderPos {
			result[i] = 1
			incrPos = 1
			borderPos = i
		} else {
			result[i] = 1 + result[incrPos]
			incrPos++
		}
	}
	return result
}

File: src/main/c/g0301_0400/s0338_counting_bits/Solution.c
// #Easy #Dynamic_Programming #Bit_Manipulation #Udemy_Bit_Manipulation
// #Big_O_Time_O(num)_Space_O(num) #2024_11_07_Time_0_ms_(100.00%)_Space_15_MB_(36.75%)

#include <stdio.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* countBits(int num, int* returnSize) {
    // Allocate memory for the result array and set the return size
    int* result = (int*)malloc((num + 1) * sizeof(int));
    *returnSize = num + 1;

    int borderPos = 1;
    int incrPos = 1;

    result[0] = 0; // Base case: 0 has zero 1-bits

    for (int i = 1; i <= num; i++) {
        // When we reach a power of 2, reset borderPos and incrPos
        if (incrPos == borderPos) {
            result[i] = 1;  // Powers of 2 have exactly one 1-bit
            incrPos = 1;
            borderPos = i;
        } else {
            result[i] = 1 + result[incrPos++];
        }
    }

    return result;
}
